### ISSUE TITLE
Tracking Issue for `[lints]` table RFC 3389

### ISSUE THREAD

**Issue by ehuss** [rust-lang/cargo#12115]
### Summary

RFC: [#3389](https://github.com/rust-lang/rfcs/pull/3389)
Implementation: #12148, #12168, #12174, #12174
Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#lints
Issues: https://github.com/rust-lang/cargo/labels/Z-lints-table

This feature adds a `[lints]` table to `Cargo.toml` to configure reporting levels for rustc and other tool lints.

See also #3591, #5034

### Testing instructions

Requires
- nightly: nightly-2023-05-25 or newer
- ~~beta~~
- ~~stable~~

Steps:

1. Switch from attributes to `[lints]` table per the [unstable reference](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#lints)
  - `-Dwarnings` should likely still be passed in explicitly so local experimentation doesn't get blocked on perfection
  - CI jobs passing in `-Adeprecations` likely should still do so since users should see deprecation warnings locally to decide whether to address them
2. Verify lints in CI with a supported compiler by adding the `-Zlints` flag

Debugging

- Build with `--verbose` to see what order `[lints]` and other `RUSTFLAGS` are being passed to the underlying tool

Example: #12178

### Updates Since RFC

- `cargo new` will automatically inherit the lints table (#12174)
- `profile.rustflags` precedence with `[lints]` was unspecified.  We have put made it a higher precedence.
- rustflags derived from `[lints]` table are not included in `rustc -C metadata` hash.
- `[lints]` should only apply to doctests as much as a diagnostic attributes (`#[warn(...)]`) in an `.rs` file applies to doctests
- Per-target or per-target-kind (bin vs lib) `[lints]` tables is deferred (not in original RFC)
  - Workaround: using diagnostic attributes (e.g. `#[warn]` in the target)
  - This is helpful when production code (a subset of `[lib]`, `[bin]`s in a workspace) want to be held to a higher standard than development code (especially tests)
    - e.g. clippy has an [unwrap_used](https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_used) lint but people might want that in tests.  clippy works around it today by having a `allow-unwrap-in-tests` setting
  - Since workspaces can have non-production libs and bins, it almost seems like this is more of user-defined lint settings groups and inheriting specific groups.  Or put another way, more design work is needed.
- Workspace-wide lints are deferred until they are needed.  See https://github.com/rust-lang/cargo/issues/12115#issuecomment-1572895133 for ideas
  - See also #12235
- We fingerprint all lints; we can optimize the fingerprinting later
  - See also https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133133126

### Unresolved Issues

- Are we still comfortable with our schema choice? https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133263409
- Are we still comfortable with our precedence choice? https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133265436

### Feedback

Cases that don't work well:
- Some lints are crate specific, like only wanting `missing_docs` and `missing_debug_implementations` for `[lib]`s ([comment](https://github.com/rust-lang/cargo/issues/12115#issuecomment-1570378356))
- Some lints are package and maybe crate specific, like `disallowed_methods` in #12178
- Explicit inheritance doesn't work well when you want to catch lints for example packages but want the example package to standalone from the workspace ([comment](https://github.com/rust-lang/cargo/issues/12115#issuecomment-1569944085))

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

**Comment by epage** [comment]
What do people think about having the unstable `[lints]` table only warn if its present and not enabled, rather than error?

This would make the behavior more as if we insta-stablized this feature and and a user then ran using an old version of cargo, giving an "unused field" warning.

I feel this would allow more people to try this feature out, increasing the amount of feedback we receive.  A project could configure their `[lints]` table and have their CI warnings job use a nightly.

The biggest downside is that normally `Cargo.toml` features require the unstable feature to be enabled via the `cargo-features` field which wouldn't work here because that would also require a nightly compiler.  Instead, this would be enabled via `-Zlints`.  We would need to strip the `[lints]` table on publish.   This is a large enough break from convention that I hesitate.

As an anecdote, I only used sparse registry once  while its been stable but off-by-default because I regularly enough run using older versions of cargo and didn't want to figure out how to workaround the "this isn't stable" error.

**Comment by flip1995** [comment]
For future discussions, here are the links to the unresolved questions and the previous discussions:

* Are we still comfortable with our schema choice? https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133263409
* Are we still comfortable with our precedence choice? https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133265436
* Can we fingerprint only the lints for the tool being run? https://github.com/rust-lang/rfcs/pull/3389#discussion_r1133133126

(Feel free to include this in the issue body and hide/delete this comment)

**Comment by epage** [comment]
Thanks for adding in the links!

**Comment by oxalica** [comment]
> The biggest downside is that normally `Cargo.toml` features require the unstable feature to be enabled via the `cargo-features` field which wouldn't work here because that would also require a nightly compiler. Instead, this would be enabled via `-Zlints`. We would need to strip the `[lints]` table on publish. This is a large enough break from convention that I hesitate.

I think we should ignored `[lints]` defined in dependency crates by default anyway, for both stable and nightly, just like the current `--cap-lints=allow` default. Warnings inside dependencies may not be solvable by end-users.

**Comment by roblabla** [comment]
A problem with erroring upon encountering the lint table without the associated feature flag is that it'll prevent users targeting a low MSRV from using the feature as soon as it gets stabilised.

**Comment by zblach** [comment]
> One thing to note here is that new lints added to clippy nightly can then suddenly break your CI (unless you pin the nightly).
> [...]
> Perhaps pinning the nightly version is a good enough workaround though.

If you're not pinning versions in CI, you're asking for this potential trouble anyway.

**Comment by epage** [comment]
Looking at #9732, it appears we have precedence for treating these as warnings, rather than errors.

**Comment by epage** [comment]
The motivation for the unstable `package.rust-version` using a warning instead of an error: https://github.com/rust-lang/cargo/pull/8037#issuecomment-763240531

**Comment by Nemo157** [comment]
It makes a lot of sense that when an unstable feature is not enabled then Cargo should mostly act as if it did not exist (other than maybe adding a note that there's a feature not enabled), I think the times when there had been errors added specifically for new features that were previously only warnings were a mistake.

**Comment by epage** [comment]
> I think the times when there had been errors added specifically for new features that were previously only warnings were a mistake. 

If its something like weak/namespaced features, there isn't always a way to map it to the old behavior, so an error makes sense.

**Comment by Nemo157** [comment]
It depends on what the previous behavior of cargo was, if without the feature implemented it was an error then it should remain an error (but could get a better message), if it was a warning it should remain a warning.

**Comment by kornelski** [comment]
Currently there's no mechanism for controlling Cargo's warnings. Could the `[lints]` be used to silence Cargo's own warnings too? Like [this one](https://github.com/rust-lang/cargo/issues/8264)?

**Comment by epage** [comment]
Yes, one of the future possibilities listed in the RFC is for cargo to have configurable lints.

**Comment by ehuss** [comment]
> What do people think about having the unstable `[lints]` table only warn if its present and not enabled, rather than error?

I think as long as it still gives a warning on stable, then that should probably be fine. I think we discussed the problem of if the schema changes, it shouldn't break older versions. That might mean doing something like parsing to `toml::Value`, and then conditionally parsing to the actual structure (if `-Zlints` is specified).

**Comment by epage** [comment]
So #12148 makes `cargo` warn if a `[lints]` table exists but use of the `[lints]` table still requires `cargo-features = ["lints"]`.  This means once `lints` is stabilized, someone can use it without bumping their MSRV.  However, it doesn't help with people using it before stabilization.  I guess someone could have their CI job inject `cargo-features = ["lints"]` directly into every `Cargo.toml`.

In some ways, it would be nice to drop the need for `cargo-features = ["lints"]` so people can use a mixture of stable/nightly on their project but I feel like thats going too far.  In rust code, users can put nightly features behind a feature flag but there isn't a way to do that with manifests.

Thought I'd post here in case people have suggestions on how to make this easier to test once its merged.

**Comment by Nemo157** [comment]
I can't find the discussion now, but I'm sure there was some about how you should be able to enable features that are used in `Cargo.toml` via `.cargo/config`. That would allow something like `cargo check --config 'cargo-features=["lints"]'` to be used.

**Comment by epage** [comment]
#9926 and #11552

> The team decided not to have a general purpose way enabling features via configuration (.cargo/config.toml). Instead, we will enable -Z flag for an unstable Cargo syntax case by case.

... huh, I was there for that discussion but don't remember it.

@ehuss think it would be appropriate for me to add a `-Z` for this feature?  Would I "or" them together or exclusively use `-Z`?

**Comment by ehuss** [comment]
I think using just `-Z` would be fine, especially since you can also enable it via the `[unstable]` config table. But if you want to use both `-Z` and `cargo-features`, that would work too, whichever you prefer.

I think for `[lints]`, things are a little more flexible since it *usually* isn't load-bearing. The only exception is when lowering a deny-by-default lint. I think that should be very rare (at least with rustc, clippy may be a different story). Perhaps a warning in the unstable docs would be sufficient to let people be aware of the backwards compatibility of that scenario?

**Comment by mcandre** [comment]
This looks very cool.

Excited for it to land in stable.

**Comment by jplatte** [comment]
I just tried this out and one thing I found is that I'd really prefer to have this configuration not require an opt-in on a per-crate basis. This is for two reasons:

- It is hard to check whether all crates have opted into the workspace lints
- When a workspace contains example crates, you want those to be copy-pastable as a standalone thing, so using `workspace = true` for anything is generally not desireable. But of course you still want to enforce a good code style for them.

**Comment by epage** [comment]
Thanks for trying this out!

Any other thoughts on the feature, including how the schema works out for you?

> I just tried this out and one thing I found is that I'd really prefer to have this configuration not require an opt-in on a per-crate basis in a setup with workspace-wide lints (IMHO the selling point of this feature, as you can just use #![warn()] for single-crate lint configuration). 

Right now, the opt-in is per-package, allowing it to cover all crates within a package (lib, bins, tests, examples, benches).  That is already a huge gain.  Whether the workspace side of this is a gain or not is dependent on how often people use "example" packages that want the feature without explicit inheritance (e.g. I don't think this will be a problem for any of the projects I touch).

Part of the concern sounds like a general concern with workspace inheritance (examples packages being able to be pulled out with non-essential fields, ensuring everything opts-in).

#12069 will help make sure packages opt-in by making it the default behavior (granted, it might make creating new examples more annoying).

It might also be worth opening an issue for exploring the wider question of implicit inheritance.  Personally, I don't think `[lints]` should try to solve implicit inheritance in a one-off way or that this has enough impact to block `[lints]` from being stabilized.  Having an issue where we can start to look for patterns in workflows can help us better understand where we should take workspace inheritance more generally. Alternatively, we could say that this is a workspace-specific field like `[patch]` but there are differences, like that `[patch]` directly modifies something that is exclusively workspace wide (resolver) and we left the door open in our design for packages to override behavior.  For example, the disallowed functions clippy lint is being used by the `cargo` lib crate to prevent bugs but that shouldn't be universally applied to all other crates in the package or workspace.  We'll need to do workarounds to use `[lints]` for now but allowing overrides could make this nicer.

**Comment by epage** [comment]
Thanks for providing the feedback!

I've recorded your workflow (transition periods when making lints stricter).  A potential workaround is to leverage features like github's SARIF support.  In theory, SARIF reporting allows you to prevent your backlog of lints from growing while people touching the relevant section of code can help reduce the backlog, rather than making it dedicated tasks.  The downside / upside is people will see the full list of warnings during local development.

FYI we originally deferred this in the RFC to future possibilities to keep the RFC more focused so we could get it through faster.

**Comment by chrisburel** [comment]
Is it possible to define different lint levels depending on tool that is being run? For instance, I like to see different lint levels in my editor (as returned by `rust-analyzer`) than I do in my command line output (via `cargo check` or `cargo clippy`). I don't want the command-line output to spew a ton of warnings that most developers are going to ignore. But the in-editor experience is different, because the signal/noise ratio is better.
I accomplish this today by specifying my lint levels with, for example, `[build] rustflags = "-Dwarnings -Adead-code" "-Aetc..."` in my `.cargo/config.toml` file, then specify `extraEnv: {"RUSTFLAGS": "-Wwarnings"}` as arguments to the `check` command in my editor's configuration for `rust-analyzer`. The environment variable overrides the setting from the `.cargo/config.toml` file, so in my editor I see warnings for `dead_code`, but when I build I don't see any warnings about it.
I tried a bit to make this work while using `-Zlints`, but it seems like I would have to use `extraArgs` to explicitly turn each `-A` lint back into a `-W` lint. Using the `RUSTFLAGS` environment variable doesn't work as a general purpose override (which I'm guessing is by design).

**Comment by epage** [comment]
>  Using the RUSTFLAGS environment variable doesn't work as a general purpose override (which I'm guessing is by design).

Per the RFC, RUSTFLAGS should win out over `[lints]` (ie come last on the command-line).  We [have a test specifically for this](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/lints.rs#L417).

If you build with `--verbose`, you should see the `rustc` or `clippy` command being run and verify what order the flags are passed down. 

If you could provide more complete reproduction steps then I can look into this.

**Comment by epage** [comment]
Oh right, iiuc `warnings` is a special group that is made up of everything that is *currently* a level=warn, so `-Dwarnings` moves them out of the group and so you can't reference them again as `warnings`.  So `-Wwarnings` is a no-op.

But I thought you had this working this way before and it should have acted the same...

**Comment by chrisburel** [comment]
What I did before was use `.cargo/config.toml` to define 
```toml
[build]
rustflags = ["-Ddead-code", # and other lints here]
```

According to [Cargo's documentation on `build.rustflags`](https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags), it has a search order:
1. `CARGO_ENCODED_RUSTFLAGS` environment variable.
2. `RUSTFLAGS` environment variable.
3. All matching `target.<triple>.rustflags` and `target.<cfg>.rustflags` config entries joined together.
4. `build.rustflags` config value.

The first one found wins, completely overriding any further rustflags. Because I set `RUSTFLAGS` in the environment that `rust-analyzer` sets up, it wins, so the `-Ddead-code` flag from my `.cargo/config.toml` isn't used at all. The flags are replaced, instead of using later flags to overrule previously specified ones.

**Comment by epage** [comment]
The cargo team hopes to use this feature to add a `[lints.cargo]` table.  That will only apply to packages though and we likely will want warnings that apply to the workspace.  This reminds me of the problem that the workspace does not have an `edition` which causes confusion when regarding `resolver`.

One option is for us to say `workspace.lints` is also for workspace-level lints.  We'll jsut have to make sure workspace-wide lints are distinct from package lints.

**Comment by lopopolo** [comment]
I love seeing config converging into `Cargo.toml`. This will help to address some papercuts when working in workspaces with many crates.

I know it is out of scope for this tracking issue, but mentioning it here since it seems relevant to the general idea. Other first party Rust tools have lots of configuration files of their own. I'd love `rustup`, `rustfmt`, and `clippy` to look in `Cargo.toml` for the configuration that today is found in `rust-toolchain.toml`, `rustfmt.toml`, and `clippy.toml`.

**Comment by faern** [comment]
I just found this feature and tried converting our workspace to use it. I also very much find it a papercut that I have to explicitly opt in to inheriting the workspace linting rules. We already have a lot of crates where we forgot to add the `#![deny(...)]` stuff in our top level `lib.rs` or `main.rs`. This just moves the problem to having to remember doing it in `Cargo.toml`. I'd love to be able to define so much more on a workspace level and have it enforced everywhere in that workspace.

Unrelated to this lint feature, but very much related to workspace inheritance. I'd love it if we could specify `publish = false`, `license = ...`, `edition = "2021"` etc only in the workspace. All crates in the workspace have and and are expected to have the same values for these fields. So just adding even more config that requires per-package explicit declaration is not ideal IMO.

**Comment by epage** [comment]
@faern one step forward is that with Rust 1.71.0, `cargo new` will now automatically inherit anything specified in the workspace.

**Comment by epage** [comment]
@rfcbot fcp merge

We've done a round of community testing and processed the feedback (tracking issue is updated).   We even have our own prototype for using this (#12178). The only things really left on this are [schema and precedence schemes](https://github.com/rust-lang/rfcs/blob/master/text/3389-manifest-lint.md#lint-precedence) which I felt like we were mostly settled on before so I figure a `merge` wouldn't be too controversial.  Looking at other schemas that I'm open to, they look like we could transition to them in the future if we find this isn't working. So this is a somewhat reversible decision, minus the options that aren't compatible and the restrictions of momentum.

You can find an example of this schema in #12178:
```toml
[workspace.lints.clippy]
all = { level = "allow", priority = -1 }
disallowed_methods = "warn"
```
As a reminder, we can take steps in the future [to improve precedence](https://github.com/rust-lang/rfcs/blob/master/text/3389-manifest-lint.md#help-the-user-with-priority)

---
Side note: of the other schemas, the one I'm most open to is:
```toml
[package]
lints = [
    { allow = "clippy::all" },
    { warn = "clippy::disallowed_methods" },
]
```
- Nice symmetry with RUSTFLAGS
- General simplicity
- Denser listing of lints without overhead of table-per-tool
- Leaves the door nicely open for this to be in target tables
- Downside being that if we add too much more to those inline tables, they'd become unwieldy because TOML doesn't yet support multi-line inline tables (currently slated for the 1.1 version of the spec)

(technically, its slightly tweaked from one of the documented proposals but close enough)

**Comment by Nemo157** [comment]
One thought I just had: should there be any support for third-party tools to read this data, or is it just up to them to parse `Cargo.toml` (and maybe the workspaces `Cargo.toml`) and extract it themselves. One usecase being my [`cargo-rustdoc-clippy`](https://github.com/Nemo157/dotfiles/blob/047bf00a8513321809df2d6fc356754e72dbf858/bin/cargo-rustdoc-clippy) tool which should probably be applying the configured lints automatically.

**Comment by Nemo157** [comment]
Skimming the docs now to verify that it'd be plausible to implement a parser from them, I notice that in the `lints.workspace = true` docs there is no mention of whether you can then customize them for this package&mdash;enabling more lints, overriding the level and/or priority of workspace configured lints&mdash;or whether inheriting forces you to only use the exact lints configured for the workspace.

**Comment by epage** [comment]
> One thought I just had: should there be any support for third-party tools to read this data, or is it just up to them to parse `Cargo.toml` (and maybe the workspaces `Cargo.toml`) and extract it themselves. One usecase being my [`cargo-rustdoc-clippy`](https://github.com/Nemo157/dotfiles/blob/047bf00a8513321809df2d6fc356754e72dbf858/bin/cargo-rustdoc-clippy) tool which should probably be applying the configured lints automatically.

Could you clarify what you mean by support for third-party tools to read this?  Like `cargo metadata` exposing it?  Seems like we've normally been handling that on a case-by-case basis.

Also, I'm not quite sure what this `cargo-rustod-clippy` tool is.  Looks like you are directly invoking the drivers.  Why?

**Comment by Nemo157** [comment]
> That is covered under Future Possibilities

Ah, it'd be good to have a note in the docs that it's not supported then.

> Could you clarify what you mean by support for third-party tools to read this? Like `cargo metadata` exposing it?

Exactly, something that exposes the data in an easier to use form, like a simple array of args with the inheritance and priorities already applied.



> Also, I'm not quite sure what this `cargo-rustod-clippy` tool is. Looks like you are directly invoking the drivers. Why?

It's a prototype of `cargo clippy --doc-tests` to allow linting of examples in doc comments, it has to invoke the `clippy-driver` directly since it's acting as `rustc` for `rustdoc`.

**Comment by jhpratt** [comment]
Somehow I missed this being proposed and implemented. I'm beyond happy that this is happening (and near stabilization!).

As far as I can tell, I can add this to the `Cargo.toml` of various crates and still support stable, correct? My current MSRV is 1.67 if it matters. I would, of course, be using the feature on nightly, but that's only an issue for local development and CI.

**Comment by epage** [comment]
> Edit: Playing around with this (trying to get a setup working), I'm not sure if this is a bug. In the workspace root's Cargo.toml, I have this

Yes, that is a bug.  When `workspace` is used, no other fields should be allowed

**Comment by jhpratt** [comment]
Oof, that's quite limiting. I would intuitively expect it to be overridden.

**Comment by epage** [comment]
We have that as a future possibility in the RFC: https://rust-lang.github.io/rfcs/3389-manifest-lint.html#packages-overriding-inherited-lints

**Comment by jhpratt** [comment]
I'll keep an eye open then! Either way, I'd love to see it stabilized. Thank you for your work Ed :)

**Comment by chrisburel** [comment]
I understand that the final commenting period is complete, but is it possible to successfully build a crate that contains a `[lints]` table with lints that do not exist in the version of Rust that is being used? From what I understand, due to rust-lang/rust#74722 (more specifically, rust-lang/rust#81906 which was marked as a duplicate to 74722), even specifying `-Aunknown-lints` will still fail to build if there is a lint that the current Rust doesn't know about. Because the `[lints]` table translates to flags given to `rustc`, I don't see how I can specify that I allow a lint that was stabilized in Rust 1.72, while still having successful builds with Rust 1.71. Can I specify lints differently depending on the version of Rust that is used? Is this outside the scope of this change?

**Comment by epage** [comment]
@weihanglo brought that to my attention and at least for myself, I felt that it wasn't sufficient to block this moving forward but that they could be worked on in parallel.

**Comment by extrawurst** [comment]
@weihanglo what is the expected timeline to stabilize it?

**Comment by jhpratt** [comment]
> what is the expected timeline to stabilize it?

It is stable on nightly. It will reach the stable compiler in 1.75.

**Comment by mcandre** [comment]
Dumb question, is the lints section in a Rust crate inherited by downstream projects?

I love linting, but I am wary of accidentally forcing a particular lint configuration onto my users.

**Comment by weihanglo** [comment]
If my memory hasn't rusty, then yes. Dependencies will compile with their lint configurations. However, users are not expected to see any compilation failure since Cargo uses [`--cap-lints`](https://doc.rust-lang.org/rustc/lints/levels.html#capping-lints) to stop dependency warnings from propogating to users. See relevant code below:

https://github.com/rust-lang/cargo/blob/d2f6a048529eb8e9ebc55d793abd63456c98fac2/src/cargo/core/compiler/mod.rs#L883-L894

**Comment by jwodder** [comment]
(I'm not sure if this issue is the best place to ask this; if there's a better location, please let me know.)

Does the syntax for the `[lints]` tables only allow per-lint settings, or can it also be used to, e.g., deny all warn-by-default lints — perhaps something like `lints.clippy.warn = "deny"`?

**Comment by epage** [comment]
`warn` is a lint group and lint groups are supported.  You need to make sure the `priority` is set correctly so it is correctly applied.

However, I would recommend against it.  In general, `#![deny(warnings)]` (and by extension doing it in the libs table) is generally frowned upon because
- whether you see warnings is dependent on the rust version used so someone using a different Rust version than you might not be able to build your project
- it means your code always has to be "perfect" even when you are experimenting on something

I would instead recommend looking at #8424.

**Comment by flip1995** [comment]
> `warn` is a lint group and lint groups are supported.

Small correction: `warn` is not a lint group in Clippy. But `all` is, which contains all warn-by-default lints. What epage wrote still applies though: `#![deny(warnings)]` and by extension `#![deny(clippy::all)]` is an [anti-pattern](https://rust-unofficial.github.io/patterns/anti_patterns/deny-warnings.html). So yo deny warn-by-default Clippy lints, you should use the same alternatives as listed in the linked documentation.

**Comment by jwodder** [comment]
> * whether you see warnings is dependent on the rust version used so someone using a different Rust version than you might not be able to build your project

Just to be clear on this point: Does the `[lints]` table have an effect when another user builds a crate that uses mine or installs my binary crate?  Or does it only take effect when building my crate directly (e.g., in a clone of the source repository)?  Does this differ between the rustc lints and the clippy lints?

**Comment by MarijnS95** [comment]
> * `cargo new` will automatically inherit the lints table

Is there (going to be) a way to force workspace lints on all member crates? Having to add `[lints] workspace = true` to every `Cargo.toml` in my workspace is not only verbose, it also turns this into an opt-in (that we'd have to watch out for very carefully in PRs that add new crates) rather than being a set of rules that are unconditionally enforced on all crates in our repo / workspace.

**Comment by epage** [comment]
#12976 was just merged which added the following to the documentation

> Generally, these will only affect local development of the current package.
Cargo only applies these to the current package and not to dependencies.
As for dependents, Cargo suppresses lints from non-path dependencies with features like
[`--cap-lints`](../../rustc/lints/levels.html#capping-lints).

**Comment by epage** [comment]
> > * `cargo new` will automatically inherit the lints table
> 
> Is there (going to be) a way to force workspace lints on all member crates? Having to add `[lints] workspace = true` to every `Cargo.toml` in my workspace is not only verbose, it also turns this into an opt-in (that we'd have to watch out for very carefully in PRs that add new crates) rather than being a set of rules that are unconditionally enforced on all crates in our repo / workspace.

#12208 is our general issue for considering implicit inheritance.  If we should consider piece-meal implicit inheritance, I think that should also be discussed there.
